### PR TITLE
Adds OData query parameters to `$count` endpoints

### PR DESCRIPTION
**Pull request by irvinesunday**
This PR:

- Fixes https://github.com/microsoft/OpenAPI.NET.OData/issues/313
- Appends OData query parameters to `$count` endpoints.
- Updates tests to validate the above.
- Updates integration test files.
- Updates release notes.

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs [hunk 1]
```diff
@@ -36,6 +39,16 @@ protected override void Initialize(ODataContext context, ODataPath path)
             int count = path.Segments.Count;
             if(count >= SecondLastSegmentIndex)
                 LastSecondSegment = path.Segments.ElementAt(count - SecondLastSegmentIndex);
+
+            if (LastSecondSegment is ODataNavigationSourceSegment sourceSegment)
+            {
+                annotatable = sourceSegment.NavigationSource as IEdmEntitySet;
+                annotatable ??= sourceSegment.NavigationSource as IEdmSingleton;
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:46]:
```suggestion
                annotatable = sourceSegment.NavigationSource as IEdmEntitySet ?? sourceSegment.NavigationSource as IEdmSingleton;
```

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs [hunk 2]
```diff
@@ -75,19 +88,85 @@ protected override void SetResponses(OpenApiOperation operation)
...
+                    operation.Parameters.Add(parameter);
+                }
             }
-            else if (LastSecondSegment is ODataNavigationPropertySegment navigationPropertySegment)
+            else
             {
-                annotatable = navigationPropertySegment.NavigationProperty;
+                // EntitySet or Collection-Valued Navigation Property
+
+                OpenApiParameter parameter = Context.CreateTop(annotatable);
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:115]: > The /$count path suffix identifies the integer count of records in the collection and SHOULD NOT be combined with the system query options  [$top](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionstopandskip), [$skip](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionstopandskip), [$orderby](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionorderby), [$expand](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionexpand), and [$format](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_SystemQueryOptionformat). The count MUST NOT be affected by $top, $skip, $orderby, or $expand.

Please remove skip, top, search, expand, count, select, orderby

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs [hunk 3]
```diff
@@ -75,19 +88,85 @@ protected override void SetResponses(OpenApiOperation operation)
             base.SetResponses(operation);
         }
 
-        protected override void AppendCustomParameters(OpenApiOperation operation)
+        /// <inheritdoc/>
+        protected override void SetParameters(OpenApiOperation operation)
         {
-            IEdmVocabularyAnnotatable annotatable = null;            
-            if (LastSecondSegment is ODataNavigationSourceSegment sourceSegment)
+            base.SetParameters(operation);
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:94]: this should also support custom defined request headers like Consistency Level defined on the parent.

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs (hunk 3 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:94]: We already add custom header here: https://github.com/microsoft/OpenAPI.NET.OData/blob/e7bad1197d1469d2cbffaebdefae470fb7cc7043/src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs#L103-L106

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs (hunk 2 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:115]: That means we only create `$filter` and `$search` and only if the parent endpoint is a collection-valued nav. prop. or an entity set?

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs (hunk 1 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs]: Getting an error here: _Operator '??' cannot be applied to operands of type IEdmEntitySet and IEdmSingleton_. Reverting to original code.

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs (hunk 2 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs]: correct, I believe the raw count path segment will only ever get added after those anyway with the code we have today.

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs (hunk 1 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:46]: could we use a common parent interface instead? like IAnnotable or something similar?

Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs]: I have opted to do away with the `IEdmSingleton` cast. We are not using it anyway

**Review by baywet — APPROVED** [review]

**Pull request merged**